[FLINK-6900] [metrics] Limit size of metric name components

zentol opened the pull request:
This PR modifies the `ScheduledDropwizardReporter` to limit the size of every metric name component to 80 characters, with the same reasoning as #4109.

greghogan commented:
@zentol since this has not yet been reviewed I'll chance a question: is this needed in addition to #4109?

zentol commented:
yes it is still required, but this PR does address a similar issue.

#4109 limits the size of the operator name in the metric identifier. This was a problem for all reporters, because a 200+ character name just isn't manageable.

This PR limits the size of all components of the metric identifier for DropwizardReporters, as the backends of several subclasses store metrics in files, with each component being one directory., like "taskmanager/abcde/job/myjob/task/mytask". Since they are used as names for directories they mustn't exceed a certain size (commonly 255). While technically a value close to 255 would suffice, i figure that anything above 80 characters isn't really well manageable as well.

greghogan commented:
What can cause such a long metric identifier? It seems risky to truncate the full identifier which could even completely remove the base name.

zentol commented:
We don't truncate the full identifier, but each individual component before assembling the final identifier.

I.e. `a.b.<reallyReallyLongSection>.c.d` would become `a.b.<notSoLongSection>.c.d`.

The primary cause of this currently are the names of WindowOperators or long task chains; but given that the sections are partially controlled by the user there may be more cases.

greghogan commented:
That sounds reasonable. Can we add a warning as in #4109 and replace `80` with a constant (I see just now in `TaskMetricGroup.java` that `80` is hard-coded in the log string rather than using the constant)?

zentol commented:
sure, i can do that.

NicoK commented:
just for clarification, `a.b.<notSoLongSection>.c.d` would then be stored in `a/b/<notSoLongSection>/c/d` so that each component/file/directory is not larger than 80?

zentol commented:
@NicoK yes, that's how they would be stored in some backends.

NicoK commented:
@zentol then +1 after addressing @greghogan's comments (adding a warning + using a constant for the `80`)

zentol commented:
I've update the PR.
* extended the change to cover the StatsDReporter
* added the warning as requested
* moved the limit into a configurable field
* updated documentation
* added tests

NicoK commented on flink-metrics/flink-metrics-dropwizard/src/main/java/org/apache/flink/dropwizard/ScheduledDropwizardReporter.java:
```diff
@@ -184,7 +188,10 @@ public void notifyOfRemovedMetric(Metric metric, String metricName, MetricGroup
+		if (metricName.length() > maxComponentLength) {
+			log.warn("The metric name component {} exceeded the {} characters length limit and was truncated.", metricName, maxComponentLength);
+		}
+		final int strLen = Math.min(metricName.length(), maxComponentLength);
```
You actually don't need to call `Math.min()` anymore after you already checked the condition for the warning message. You could thus assign `strLen` yourself.

NicoK commented on flink-metrics/flink-metrics-statsd/src/test/java/org/apache/flink/metrics/statsd/StatsDReporterTest.java:
```diff
@@ -60,6 +60,20 @@
+		
+		reporter.open(config);
+		
+		assertEquals("0123456789", reporter.filterCharacters("0123456789DEADBEEF"));
```
...and the additional tests

zentol replied:
What about the tests?

NicoK replied:
hmm, seems something was lost during the transfer :(

iirc, I wanted to ask whether it makes sense to also test that things like `a.b.0123456789DEADBEEF.c` are properly truncated to `a.b.0123456789.c` for the whole metric name

NicoK commented on flink-metrics/flink-metrics-statsd/src/main/java/org/apache/flink/metrics/statsd/StatsDReporter.java:
```diff
@@ -193,7 +198,10 @@ private void send(final String name, final String value) {
+		if (input.length() > maxComponentLength) {
+			log.warn("The metric name component {} exceeded the {} characters length limit and was truncated.", input, maxComponentLength);
+		}
+		final int strLen = Math.min(input.length(), maxComponentLength);
```
same here about the `Math.min`

zentol commented on docs/monitoring/metrics.md:
```diff
@@ -376,6 +376,7 @@ Parameters:
 - `dmax` - hard limit for how long an old metric should be retained
 - `ttl` - time-to-live for transmitted UDP packets
 - `addressingMode` - UDP addressing mode to use (UNICAST/MULTICAST)
+- `maxComponentLength` - limits the size of each scope component
```
"length of each scope component" would be better, I don't think we us "name of scope component" anywhere in the docs.

greghogan replied:
Alright, works for me.
